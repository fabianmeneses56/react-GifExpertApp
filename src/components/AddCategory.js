import React, { useState } from "react";
import PropTypes from "prop-types";

export const AddCategory = ({ setcategories }) => {
  const [inputValue, setinputValue] = useState("");

  const handleInputChange = (e) => {
    setinputValue(e.target.value);

    console.log('handleInput llamado')
  };

  const handleSubmit = (e) => {
    /* prevenir que la pagina se actualice por navegador */
    e.preventDefault();
    console.log('handle submit',inputValue);
    /* validacion de que hallna mas de 2 letras */
    if (inputValue.trim().length > 2) {
      setcategories((cats) => [inputValue, ...cats]);
      setinputValue("");
    }
  };

  return (
    <form onSubmit={handleSubmit}>
      <p>{ inputValue }</p>
      <input type="text" value={inputValue} onChange={handleInputChange} />
    </form>
  );
};

AddCategory.propTypes = {
  setcategories: PropTypes.func.isRequired,
};
