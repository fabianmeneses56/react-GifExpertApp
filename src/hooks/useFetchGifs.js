import React, { useEffect, useState } from "react";
import { getGifs } from "../helpers/GetGifs";

export const useFetchGifs = (category) => {
  const [state, setstate] = useState({
    data: [],
    loading: true,
  });

  useEffect(() => {
    getGifs(category).then((imgs) => {
      //setTimeout(() => {
         /*  console.log(imgs) */
        setstate({
          data: imgs,
          loading: false,
        });
      //}, 3000);
    });
  }, [category]);

  /* devuelve lo que esta adentro despues del tiempo establecido */
  /*    setTimeout( ()=>{
        setstate({
            data:[1,2,3,4],
            loading: false
        })
    },3000); */

  return state; // data y loading
};
