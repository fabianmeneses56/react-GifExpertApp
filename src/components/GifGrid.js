import React from "react";
import { useFetchGifs } from "../hooks/useFetchGifs";
import { GifGridItem } from "./GifGridItem";
import PropTypes from 'prop-types'

// EN ESTE SE IMPLEMENTO UN CUSTOM HOOK QUE HACE EL TRABAJO PESADO
export const GifGrid = ({ category }) => {
                    /*   const [images, setimages] = useState([]); */

                    /* use effect es como el component didmount */
  const { data:images,loading } = useFetchGifs( category );
                    /*  useEffect(() => {
                        //getGifs(); -> como ya no tengo el getgifs lo importo
                        getGifs(category).then(setimages);
                        //.then(imgs=>setimages(imgs))
                    }, [category]); */

                    /*  getGifs(); */

  return (
    <>
      <h3 className="animate__animated animate__fadeIn">{category}</h3>

      {loading && <p className="animate__animated animate__flash">cargando</p>}
     
      <div className="card-grid">
        <ol className="ol-grid">
          {images.map((img) => (
            <GifGridItem
              key={img.id}
              {...img}
              //img={img}
            />
          ))}
        </ol>
      </div>
    </>
  );
};
GifGrid.propTypes={
  category:PropTypes.string.isRequired
}