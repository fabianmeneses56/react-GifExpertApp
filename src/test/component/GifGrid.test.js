import React from 'react';
import {shallow}  from 'enzyme';
import { GifGrid } from '../../components/GifGrid';
import { useFetchGifs } from '../../hooks/useFetchGifs';
jest.mock('../../hooks/useFetchGifs');

describe('Pruebas en el componente <GifGrid/>', () => {

    const category="harry potter";


    test('debes mostrar el componente correctamente', () => {

        useFetchGifs.mockReturnValue({
            data: [],
            loading: true,
        })
        const wrapper=shallow(<GifGrid category={ category }/>)
        expect( wrapper ).toMatchSnapshot();
    })

    test('debe de mostrar items cuando se cargan imagenes useFetchGifs ', () => {
        //Mock
        const gifs=[{
            id:'ABC',
            url:'http/cualquier/cosa',
            title:'cualquier cosa'
        }]

        useFetchGifs.mockReturnValue({
            data: gifs,
            loading: false,
        })

        const wrapper=shallow(<GifGrid category={ category }/>)

        expect( wrapper ).toMatchSnapshot();

        //evaluar si existen componentes

        expect( wrapper.find('p').exists() ).toBe(false);
        expect( wrapper.find('GifGridItem').length ).toBe(gifs.length)
    })
    
    
    
})
