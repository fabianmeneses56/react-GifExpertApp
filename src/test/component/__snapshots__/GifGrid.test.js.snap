// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el componente <GifGrid/> debe de mostrar items cuando se cargan imagenes useFetchGifs  1`] = `
<Fragment>
  <h3
    className="animate__animated animate__fadeIn"
  >
    harry potter
  </h3>
  <div
    className="card-grid"
  >
    <ol
      className="ol-grid"
    >
      <GifGridItem
        id="ABC"
        key="ABC"
        title="cualquier cosa"
        url="http/cualquier/cosa"
      />
    </ol>
  </div>
</Fragment>
`;

exports[`Pruebas en el componente <GifGrid/> debes mostrar el componente correctamente 1`] = `
<Fragment>
  <h3
    className="animate__animated animate__fadeIn"
  >
    harry potter
  </h3>
  <p
    className="animate__animated animate__flash"
  >
    cargando
  </p>
  <div
    className="card-grid"
  >
    <ol
      className="ol-grid"
    />
  </div>
</Fragment>
`;
